File: java源码/offer/Test16.java
package offer;

/**
 * 剑指offer-题16：反转链表
 * 
 * 输入一个链表的头结点，反转该链表并输出反转后链表的头结点。
 */

public class Test16 {
	public ListNode ReverseList(ListNode head) {
		ListNode pNode = head;
		ListNode pPrev = null;
		ListNode pNest = null;
		ListNode pReversedHead = null;
		while (pNode != null) {
			pNest = pNode.next;
			if (pNest == null) {
				pReversedHead = pNode;
			}
			pNode.next = pPrev;
			pPrev = pNode;
			pNode = pNest;
		}
		return pReversedHead;
	}

	public static void main(String[] args) {
		Test16 test16 = new Test16();
		ListNode a = new ListNode(1);
		ListNode b = new ListNode(2);
		ListNode c = new ListNode(3);
		ListNode d = new ListNode(4);
		ListNode e = new ListNode(5);
		ListNode f = new ListNode(6);
		a.next = b;
		b.next = c;
		c.next = d;
		d.next = e;
		e.next = f;
		ListNode node = test16.ReverseList(a);
		while (node != null) {
			System.out.print(node.val + ", ");
			node = node.next;
		}
	}

}

File: python3源码/Test16.py
# -*- coding:utf-8 -*-

# 剑指offer-题16：反转链表

# 输入一个链表的头结点，反转该链表并输出反转后链表的头结点。


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # 返回ListNode
    def ReverseList(self, pHead):
        pNode = pHead
        pPrev = None
        pReversedHead = None
        while pNode is not None:
            pNext = pNode.next
            if pNext is None:
                pReversedHead = pNode
            pNode.next = pPrev
            pPrev = pNode
            pNode = pNext
        return pReversedHead

solution = Solution()
a = ListNode(1)
b = ListNode(2)
c = ListNode(3)
d = ListNode(4)
e = ListNode(5)
f = ListNode(6)
a.next = b
b.next = c
c.next = d
d.next = e
e.next = f
node = solution.ReverseList(a)
while node is not None:
    print(node.val, end=", ")
    node = node.next

